sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/errorhandling/ErrorHandler.java

```diff
@@ -0,0 +1,222 @@
+    @Override
+    public @Nullable OutputT getOutput() {
+      if (!this.isClosed()) {
+        throw new IllegalStateException(
+            "ErrorHandler must be finalized before the output can be returned");
+      }
+      return sinkOutput;
+    }
+
+    @Override
+    public void close() {
+      closed = true;
+      PCollection<ErrorT> flattened;
+      if (errorCollections.isEmpty()) {
+        LOG.warn("Empty list of error pcollections passed to ErrorHandler.");
+        flattened = pipeline.apply(Create.empty(coder));
+      } else {
+        flattened = PCollectionList.of(errorCollections).apply(Flatten.pCollections());
+      }
+      LOG.debug(
+          "{} error collections are being sent to {}",
+          errorCollections.size(),
+          sinkTransform.getName());
+      String sinkTransformName = sinkTransform.getName();
+      sinkOutput =
+          flattened
+              .apply(
+                  "Record Error Metrics to " + sinkTransformName,
+                  new WriteErrorMetrics<ErrorT>(sinkTransformName))
+              .apply(
+                  "Write to error Sink",
+                  sinkTransform.addAnnotation(
+                      "FeatureMetric", "ErrorHandler".getBytes(StandardCharsets.UTF_8)));
+    }
+
+    public static class WriteErrorMetrics<ErrorT>
+        extends PTransform<PCollection<ErrorT>, PCollection<ErrorT>> {
+
+      private final Counter errorCounter;
+
+      public WriteErrorMetrics(String sinkTransformName) {
+        errorCounter = Metrics.counter("ErrorMetrics", sinkTransformName + "-input");
+      }
+
+      @Override
+      public PCollection<ErrorT> expand(PCollection<ErrorT> input) {
+        return input.apply(ParDo.of(new CountErrors<ErrorT>(errorCounter)));
+      }
+
+      public static class CountErrors<ErrorT> extends DoFn<ErrorT, ErrorT> {
+
+        private final Counter errorCounter;
+
+        public CountErrors(Counter errorCounter) {
+          this.errorCounter = errorCounter;
+        }
+
+        @ProcessElement
+        public void processElement(@Element ErrorT error, OutputReceiver<ErrorT> receiver) {
+          errorCounter.inc();
+          receiver.output(error);
+        }
+      }
+    }
+  }
+
+  class BadRecordErrorHandler<OutputT extends POutput>
+      extends PTransformErrorHandler<BadRecord, OutputT> {
+
+    /** Constructs a new ErrorHandler for handling BadRecords. */
+    @Internal
+    public BadRecordErrorHandler(
+        PTransform<PCollection<BadRecord>, OutputT> sinkTransform, Pipeline pipeline) {
+      super(sinkTransform, pipeline, BadRecord.getCoder(pipeline));
+    }
+  }
+
+  /**
+   * A default, placeholder error handler that exists to allow usage of .addErrorCollection()
+   * without effects. This enables more simple codepaths without checking for whether the user
+   * configured an error handler or not.
+   */
+  @Internal
+  class NoOpErrorHandler<ErrorT, OutputT extends POutput> implements ErrorHandler<ErrorT, OutputT> {
```

Comment:
NoOp seems to imply that we do nothing with the errors (like /dev/null). What we want here is something that we can call addErrorCollection on when we know that the passed error collection will be empty, but should never be used in as an actual error handler. In particular, if there happens to be something in those collections, that's a bug. (Should this be asserted? Actually, in this case it's fine to call close() and all. It's a NoneShallPassErrorHandler.)

I just can't shake the feeling that this'd all be so much easier if we finish pushing error handling into ParDo itself...

Reply:
I'll rename this. I agree that I'll probably want to rework this once we update the ParDo itself, but that will take quite a lot of design that we don't have in place yet.